### PR TITLE
Adaptive solver / mechanism auto-reduction feature

### PR DESCRIPTION
**Pull request by jimmielin**
Hi all,

I'm opening up this PR for the eventual merge of the auto-reduce feature into KPP 3.0.0, which @msl3v and I have been working on.

For some reason GitHub shows a few files changed that don't belong to this feature. It'll be necessary to merge this with a little manual work. As far as I know, the changes should be in these files only:

- int/rosenbrock_autoreduce.def
- int/rosenbrock_autoreduce.f90
- src/code.h (support `LOGICAL`)
- src/code_f90.c (support `LOGICAL`)
- src/gen.c
- src/scan.l
- src/scan.y
- src/scanner.c

To enable this, run `kpp` with an input file that has `#AUTOREDUCE on`. The `#INTEGRATOR rosenbrock_autoreduce` must be selected.

To use AR at runtime, `ICNTRL(8) = .true.` needs to be set. Two thresholding methods are supported:

1. `RCTNRL(8)` sets an absolute threshold in molec cm-3
2. or, `ICNTRL(10) /= 0` chooses a target species ID to base AR threshold on; `RCNTRL(10)` is the factor that is multiplied to the target species' `max(P, L)` to be used as the threshold

The # for 8, 10 here may need to be changed to be standardized across integrators during the merge, noting that GEOS-Chem uses these as well and `fullchem_mod.F90` needs to be changed accordingly.

### CODE REVIEW

**Comment by jimmielin** [comment]
Hi Rolf, I agree with this course of action. I think once `cleanup_int` and `docs` are merged into `dev`, this PR's changes will be clearer (a lot of the conflicts seem to be with the docs, probably because I rebased `autoreduce` over some commit in `dev`).

Then we can see if `autoreduce` can be merged in directly into `dev`. I don't see why it would negatively affect `dev` since we took care in implementing this into a separate integrator, `rosenbrock_autoreduce.f90`, so the sooner it's done the better before everything starts to diverge too far. Please let me know what you think!

**Comment by RolfSander** [comment]
Unless @yantosca or @msl3v have any objections, I'm happy to merge
cleanup_int and docs into dev now. See:

https://github.com/KineticPreProcessor/KPP/pull/44

Let's see if there will be any remaining conflicts...

**Comment by yantosca** [comment]
Once PR #47 is approved then we can go ahead and merge cleanup_int into dev.

**Comment by RolfSander** [comment]
We have now merged both `cleanup_int` and `docs` into `dev` now.

@jimmielin: You can merge `dev` now into your `autoreduce` branch and check for any potential remaining conflicts there.

**Comment by jimmielin** [comment]
Thanks. I've attempted a merge of `dev` into `autoreduce`. I've cleared all the git conflicts, but I need to make sure the AR functionality is still intact and that it builds ok. GitHub still seems to show some unrelated files in `doc` being updated for some reason. I'll report back...

**Comment by jimmielin** [comment]
Thanks Rolf! Looks like the following changes are needed for GEOS-Chem once this is merged:

* `#UPPERCASEF90` isn't enabled in the 13.4.0 version I have yet
* `#INCLUDE atoms` in `fullchem.eqn` needs to be changed to `#INCLUDE atoms.kpp`.

I'm building this now and testing out the new generated code inside GEOS-Chem. Once I complete a full run, it looks like the merge should work. I don't know why the CI tests are failing (I haven't touched the other integrators at all and #AUTOREDUCE is off by default).

**Comment by jimmielin** [comment]
Thanks @yantosca, fixed. Let's see if the CI tests pass. Also I had to make a fix to `OHreact_parser.py` - I'll make a separate PR for that over at the GEOS-Chem repo since it's a backwards compatible change.

**Comment by yantosca** [comment]
@jimmielin -- if you want to run the C-I tests locally, do this:

```console
cd $KPP_HOME/.ci-pipelines
./ci-manual-testing-script.sh
```
and then that should do the same thing as the C-I tests on Azure.  (It won't run them in the container but that's good enough for finding bugs).

**Comment by yantosca** [comment]
and then `./ci-manual-cleanup-script.sh` to remove all of the generated output from the local C-I tests.

**Comment by jimmielin** [comment]
Thanks Bob! That'll be useful, I'm still running into a segfault in the CI test. I'll debug it locally.

I've just noticed that you made the fix in GEOS-Chem 14 as well. Looks like this will be necessary only for the 13.4.0 testing version I have. I'll try to run the full model with and without auto-reduction, but we are getting close to making this mergeable.

**Comment by yantosca** [comment]
Great.  What I like to do is to run the C-I tests on the branch that I'm working on, and then again on the parent branch (requires make distclean and rebuilding kpp) and then diff them to see if there are results.  (I pipe the output of ./ci-manual-testing-script.sh to log files).

**Comment by yantosca** [comment]
Hi @jimmielin, yes those variables were from the version of KPP that was kind of "hardwired" for GEOS-Chem.  They are now declared locally in GEOS-Chem in the commonIncludeVars.H file.

Also I noticed a couple of unused variable warnings in the C code:
```console
gen.c:433:5: warning: unused variable ‘flxind’ [-Wunused-variable]
  433 | int flxind[MAX_EQN];
      |     ^~~~~~
gen.c:423:8: warning: unused variable ‘j’ [-Wunused-variable]
  423 | int i, j;
      |        ^
```
Removing the unused variables will clear the warnings.

**Comment by jimmielin** [comment]
Thanks @yantosca. I've backported some changes from GC 14 into my copy of GC 13.4 so it compiles and runs. I've also just updated `rosenbrock_autoreduce`, the Rosenbrock solver with auto-reduction feature, to be in sync with other changes made in `rosenbrock.f90`. Pending checks on the output I think this is ready to merge into `dev`. Hopefully with all the changes made now, it should be painless.

**Comment by yantosca** [comment]
Thanks @jimmielin.  Feel free to merge into dev pending successful checks.

**Comment by RolfSander** [comment]
I would like to run a few tests with my MECCA model, using the code from
the autoreduce branch but with `#AUTOREDUCE OFF`. Hopefully I will get
(almost) the same results as before. I think I can do these tests until
tomorrow.

**Review by RolfSander — APPROVED** [review]
I have tested the code in the autoreduce branch with MECCA: It works
fine!

And yes, later I will certainly also try MECCA with `#AUTOREDUCE ON` :-)

When comparing the codes of dev and autoreduce directly, there were many
white-space only differences. To make the comparison easier, I made some
cosmetic changes in autoreduce.

The only remaining question is about the new files in docs/source:

```
basic-info/about.rst
creating-mechanisms/configuration_files.rst
usage-details/generating_f90_code.rst
usage-details/installation.rst
usage-details/use_custom_mech.rst
```

Maybe these are all GEOS-Chem specific and should be moved to the
GEOS-Chem documentation?

File: docs/source/basic-info/about.rst
Comment (yantosca) [inline docs/source/basic-info/about.rst:1]: I think this is the old documentation page.  Want to be sure that this doesn't clobber the updated doc.

File: docs/source/creating-mechanisms/configuration_files.rst
Comment (yantosca) [inline docs/source/creating-mechanisms/configuration_files.rst:1]: This is GEOS-Chem specific and not needed.  I have since moved this content to geos-chem.readthedocs.io.

File: docs/source/usage-details/generating_f90_code.rst
Comment (yantosca) [inline docs/source/usage-details/generating_f90_code.rst:1]: Ditto, geos-chem specific

File: docs/source/usage-details/installation.rst
Comment (yantosca) [inline docs/source/usage-details/installation.rst:1]: Also geos-chem speciifc

**Review by yantosca — APPROVED** [review]
We can remove the docs/source/basic-info folder.  Otherwise good to merge.

**Comment by RolfSander** [comment]
I noticed that autoreduce uses additional ICNTRL (8,9,10), RCNTRL
(8,10), and RSTATUS (4) elements. Can we add short descriptions of them
to 04_input_for_kpp.rst and 05_output_from_kpp.rst in
docs/source/using_kpp?

**Comment by jimmielin** [comment]
I agree, @RolfSander. I was worried that the switches' meaning would be inconsistent as well. I will move both `ICNTRL` and `RCNTRL` added to positions 12, 13, and 14, for consistency. I will push a commit shortly both here and over at GEOS-Chem.

**Comment by RolfSander** [comment]
@jimmielin: Thanks! Yes, 12, 13, and 14 are suitable positions.

@yantosca: How can I activate the updated *.rst files? Or are you doing this?

**Comment by yantosca** [comment]
@jimmielin: For now the ReadTheDocs output is being built from the docs branch but we can change that when we release.

**Comment by RolfSander** [comment]
Oh, sorry, I didn't know this. I deleted the docs branch after it was
merged into dev. I guess we'll have to recreate it (or choose another
branch to build the docs.

**Comment by yantosca** [comment]
For now I can point it to the main branch, so that the docs will be pegged to KPP 2.5.0 until we merge in the new features.

**Comment by jimmielin** [comment]
Verified that the changed version of KPP works correctly with GEOS-Chem as well over at https://github.com/jimmielin/geos-chem/tree/staging/autoreducekpp. I'm happy to merge this anytime.

**Comment by RolfSander** [comment]
Merging is fine for me now.

**Comment by yantosca** [comment]
Thanks @jimmielin!  We should merge.  Were we going to release this as an intermediate 2.6.0 version, or should we just hold off until 3.0.0?

**Comment by RolfSander** [comment]
I think we can call it 2.6.0 now. After that we can take care of the cleanup_util branch and check the documentation.

**Comment by yantosca** [comment]
Thanks.  I can do the merge & do a review of the docs before pushing a release.

**Comment by yantosca** [comment]
I've pushed the KPP 2.6.0 release and made some minor updates to the doc (such as adding the reference to @jimmielin's paper in prep).  But I had to manually create a DOI ([DOI](https://doi.org/10.5281/zenodo.6727764)) since there was a bug in the `.zenodo.json` file (now fixed in cleanup_util).

On to KPP 3.0.0!

**Comment by RolfSander** [comment]
@jimmielin: I have two questions about rosenbrock_autoreduce.f90:

- Many variables are declared as `REAL(kind=dp)`. Is this correct or
  should it be `KPP_REAL`?

- I think `cWAXPY` is never used. Can it e deleted?

**Comment by jimmielin** [comment]
Hi @RolfSander,

1. I think it can be replaced with `KPP_REAL`.
2. Yes. The version committed here has been highly optimized and is not the original implementation by @msl3v. I removed the original implementation (which used `cWAXPY` and had a much clearer flow in code...) to clean up the merge.

I'm happy to work on and submit a pull request to clean up the code and some cosmetic changes to remove trailing spaces etc which trip up `git merge --check`.

**Comment by RolfSander** [comment]
@jimmielin: Thanks!

Since we don't have the autoreduce branch anymore, I think you can put these small changes directly into dev  (if @yantosca agrees).

**Comment by yantosca** [comment]
That is fine.  We can push a 2.6.1 release, which would also fix the Auto-generation of the DOI.

**Pull request merged**
